### ISSUE TITLE
[jvm] Unexpected/Broken behavior when working with public properties in class from external Jar

### ISSUE THREAD

**Issue by davidbruce** [HaxeFoundation/haxe#10080]
So, this is a pretty odd issue that I recently stumbled upon.

I am working with [JOML](https://github.com/JOML-CI/JOML) Vectors and I have found that trying to access public properties (x,y,z,w) gives me some weird errors.

For example when working with [Vector3f](https://github.com/JOML-CI/JOML/blob/abcec058398032b62353cf28c60d98009927ff4c/src/org/joml/Vector3f.java#L52) 

`trace(vector.x)` gives me the following:
```
_ArcBallCameraDemo.ArcBallCameraDemo_Fields_$Vector3f_x@d4a2a0ad
```

`trace(cast(vector.x, Float))` gives me:
```
_ArcBallCameraDemo.ArcBallCameraDemo_Fields_$Vector3f_x incompatible with java.lang.Number
```

`trace(vector.x())` works as expected giving me:
```
0.0
```

`trace(vector.x() * -1)` gives me -0.0, which shouldn't be possible.
```
-0.0
```

`trace(-vector.x())` gives me the following with AdoptOpenJDK OpenJ9 15.0.1:
```
Error: Unable to initialize main class _ArcBallCameraDemo.ArcBallCameraDemo_Fields_
Caused by: java.lang.VerifyError: JVMVRFY012 stack shape inconsistent; class=_ArcBallCameraDemo/ArcBallCameraDemo_Fields_, method=loop()V, pc=101; Type Mismatch, argument 0 in signature haxe/jvm/Jvm.opNeg:(Ljava/lang/Object;)Ljava/lang/Object; does not match
Exception Details:
  Location:
    _ArcBallCameraDemo/ArcBallCameraDemo_Fields_.loop()V @101: JBinvokestatic
  Reason:
    Type float (current frame, stack[1]) is not assignable to 'java/lang/Object'
  Current Frame:
    bci: @101
    flags: { }
    locals: { 'org/joml/Matrix4f', 'java/nio/FloatBuffer' }
    stack: { 'haxe/jvm/Function', float }
  Stackmap Table:
    append_frame(@14,Object[#74],Object[#78])
    same_frame(@23)
    same_frame_extended(@235)
Error: Command failed with error 1
```

`trace(-vector.x())` gives me the following with AdoptOpenJDK Hotspot 15.0.0:
```
Error: Unable to initialize main class _ArcBallCameraDemo.ArcBallCameraDemo_Fields_
Caused by: java.lang.VerifyError: Bad type on operand stack
Exception Details:
  Location:
    _ArcBallCameraDemo/ArcBallCameraDemo_Fields_.loop()V @101: invokestatic
  Reason:
    Type float (current frame, stack[1]) is not assignable to 'java/lang/Object'
  Current Frame:
    bci: @101
    flags: { }
    locals: { 'org/joml/Matrix4f', 'java/nio/FloatBuffer' }
    stack: { 'haxe/jvm/Function', float }
  Bytecode:
    0000000: bb00 d659 b700 d84b 1010 b800 de4c b200
    0000010: 5eb8 00e4 9a00 d703 03b2 004b b200 4eb8
    0000020: 00e8 b200 ebb2 00ee 80b8 00f1 b200 f4b8
    0000030: 00f7 2a14 00f8 b200 fe6b 1400 ff6f 90b2
    0000040: 004b 87b2 004e 876f 9013 0101 1064 86b6
    0000050: 0105 2bb6 0109 b801 0db2 0113 c000 6ab2
    0000060: 0116 b601 1cb8 0121 b801 25bb 0127 5913
    0000070: 0129 1301 2b10 7213 012d b701 30b6 0134
    0000080: 2a03 8603 86b2 0137 7790 b601 3bb2 013e
    0000090: 90b6 0142 b201 4590 b601 48b2 0116 b601
    00000a0: 1c02 866a b201 16b6 014b 0286 6ab2 0116
    00000b0: b601 4e02 866a b601 5157 b201 54b8 00f7
    00000c0: 2a2b b601 09b8 010d b801 572a b201 16c0
    00000d0: 0159 b601 5c2b b601 09b8 010d b801 5fb2
    00000e0: 005e b801 62b8 0165 a7ff 26b1          
  Stackmap Table:
    append_frame(@14,Object[#214],Object[#224])
    same_frame(@23)
    same_frame_extended(@235)

Error: Command failed with error 1
```

I would expect to be able to just do `-vector.x` and not `-vector.x()` or `vector.x() * -1`. 

I found this issue with my code here:

https://github.com/davidbruce/haxe-lwjgl-examples/blob/9aa172bc0ef35cb5ed704646982d1818fc3bdab4/src/ArcBallCameraDemo.hx#L117

I'm getting the JOML jars from my project [here](https://github.com/davidbruce/haxe-joml) and LWJGL [here](https://github.com/davidbruce/haxe-lwjgl)

I am on OSX using Haxe version `4.2.0-rc.1+5145ddb` from the build on 2021-01-16

**Comment by Simn** [comment]
At a first glance this looks pretty wild. `_ArcBallCameraDemo.ArcBallCameraDemo_Fields_$Vector3f` suggests that we somehow have a `Vector3f` sub class of... the module fields of `ArcBallCameraDemo` or something insane like that.

I guess I'll have to inspect the involved .jar files to figure out what's going on here.

**Comment by Simn** [comment]
Feature request: Could I get a .zip file with all the stuff I need to reproduce this?

**Comment by davidbruce** [comment]
Hey Simn, I've uploaded a zip with everything you need.

You'll need to 
`haxelib dev haxe-joml haxe-joml/`
and
`haxelib dev haxe-lwjgl haxe-lwjgl/`


Zip Structure:
- haxe-joml/
  - src/
    - joml/
      - _internal/
        - jars/
          - common/ (This is where the jars you probably want to look at reside)
- haxe-lwjgl/
- haxe-lwjgl-examples/
  - joml/
    - ArcBallCameraDemo.hx  (On line 115 is a trace showing `_ArcBallCameraDemo.ArcBallCameraDemo_Fields_$Vector3f_x`)
  - output/ (premade jars for MacOS and Linux here)



[issue10080.zip](https://github.com/HaxeFoundation/haxe/files/6005219/issue10080.zip)

**Comment by Simn** [comment]
Ok I misunderstood the problem here. That `Vector3f` class has both a field and a method named `x`, which isn't really something that we support in Haxe. I'm not sure if I was aware that this is even allowed, so the .jar loader doesn't deal with it properly.

... but also, I'm not sure how it would deal with this properly. If this is a common Java pattern then I suppose we should prioritize the field over the method. We could also rename the method to `get_x` and generate the field as `var x(get, default)`, but that runs the risk of further name clashes.

**Comment by Simn** [comment]
Another option would be to add a feature that rewrites read (not call) access to a method. Something along the lines of:

```haxe
class Vector3f {
    @:native("x")
    public var _hx_x:Float;

    @:hx.read("_hx_x")
    public function x() {
        return _hx_x;
    }
}
```

This would mean that overrides still work, with the downside being that we can no longer take the closure of the method.

**Comment by davidbruce** [comment]
I agree that renaming it to `get_x` could possibly lead to name clashes, but seeing as how standard java code is camel case it might only happen in an extreme case.  

Also I apologize, but for the other solution what do you mean by `we can no longer take the closure of the method`?

Lastly, the more I look at the JOML code base the more confused I am.  There are multiple classes that have functions that do the same thing, just one is private and one is public.  So perhaps this isn't the best library to go off of for this issue.
